### ISSUE TITLE
Caps Lock light not working

### ISSUE THREAD

**Issue by miningslayer18** [jakeday/linux-surface#245]
I have an SP3 with an SP3 type cover. I'm running Ubuntu 18.04. When I press Caps Lock, it does enable, but the light does not turn on on the type cover, so there is no way of telling whether it's on or not without typing.

**Comment by qzed** [comment]
The same happens on the SB2, probably other devices are affected too?

**Comment by JackHull** [comment]
Yes, I can confirm this issue with my SP4 + type cover, running Ubuntu 18.04 LTS.

**Comment by PicoLegion** [comment]
Just noticed this problem with my SP2017 running Mint 19. For me it's only with the 4.18.7 kernel. If I switch back to 4.17.3, it functions again.

**Comment by xaviervilla** [comment]
Same on my surface book 4.17.3 on ubuntu 16.04. No big deal to me, I'm just glad it's not hardware related! Didn't expect this to be an issue listed here, just thought I would check.

**Comment by xaviervilla** [comment]
Inside of /sys/class/leds/ there are duplicate folders for the LEDs. there's input60::capslock and input2:capslock. I wonder if the capslock LED is being triggered twice when pressed. So it gives the illusion that it's not changing state, but it's actually being flipped twice. Can't think of a way to test since I don't know how to modify stuff in /sys/.

**Comment by xaviervilla** [comment]
Got it!

pressing the caps lock button changes the state of input2::capslock, but it needs to instead change input60::capslock. Not sure how to fix that though.

**Comment by battaglia01** [comment]
I have the same problem, and after some digging it seems to be complex.

Do you folks find that it works in a tty, say if you hit Ctrl-Alt-F1 and then type caps lock?

Can you type "kbd_mode" at a tty and tell me what it says?

**Comment by xaviervilla** [comment]
@battaglia01 tty does not enable LED for me either. `kbd_mode` shows:
`keyboard is in unicode (UTF-8) mode`

**Comment by battaglia01** [comment]
That's a neat way to do it. I'm kind of curious to dig into this more though.

What I find is if I type `xset -q`, and Caps Lock is on (but the light is off), I get the following:

```
$ xset -q                                                                                                                         
Keyboard Control:
  auto repeat:  on    key click percent:  0    LED mask:  00000001
  XKB indicators:
    00: Caps Lock:   on     01: Num Lock:    off    02: Scroll Lock: off
    03: Compose:     off    04: Kana:        off    05: Sleep:       off
    06: Suspend:     off    07: Mute:        off    08: Misc:        off
    09: Mail:        off    10: Charging:    off    11: Shift Lock:  off
    12: Group 2:     off    13: Mouse Keys:  off
```
So it does think caps lock is on. But then, the following command disagrees:

```
$ sudo kbdinfo gkbled                                                                                                             
scrolllock:off numlock:off capslock:off
```

So `xset` thinks the capslock indicator is on (when it isn't), but kbdinfo realizes that it isn't.

I found some partial success here: https://unix.stackexchange.com/questions/136817/caps-lock-led-not-working-on-linux-console

This gets things working on tty, although it seemed iffy. After making the changes and then resetting them, the console works now without them, even though things are exactly the way they were before. Hm...

**Comment by xaviervilla** [comment]
For me, the same thing. kdbinfo says it's off and xset thinks it's on. What I think is strange is that `ls /sys/class/leds/` shows
```
input2::capslock  input2::numlock  input2::scrolllock  input60::capslock  input60::numlock  input60::scrolllock
```
and pressing the capslock button (before my workaround script was in place) changes the state of the input2::capslock/brightness file between 1(on) and 0(off) but the LED doesn't change. Stranger still, input60::capslock/brightness does not change with the button, but when I change it manually, it does toggle the LED. IDK, I'm a noob to this lower level stuff.

**Comment by battaglia01** [comment]
Right. So what does fix it for me is the following magic command:

`sudo kbd_mode -u`

But then, some shortcuts break. Alt-left and alt-right give you tty1 and tty2 and so on. Does that happen for you?

**Comment by battaglia01** [comment]
I have the problem on 4.18.0 but not 4.16.0, fwiw.

**Comment by xaviervilla** [comment]
@battaglia01 Your command `sudo kbd_mode -u` did not fix the led for me. It did however give the weird alt+left tty shortcut though. To get back to normal mode would I use the `sudo kbd_mode -a` command?

**Comment by xaviervilla** [comment]
Okay so today I had 3 back to back occurrences of my laptop spazing out. I've been stable for a long time and haven't made any changes besides my script and runing `sudo kbd_mode -u`. Could it be related? When I resume my laptop from hibernate it doesnt have any window bars or start bars and whatever app that was open on my screen flickers erratically and shows my desktop. 

Checking `dmesg` shows something that seems like it may be peculiar. It finds a bunch of unknown surface keyboard inputs. Maybe this is normal, I just want to provide info that may help.
```
[    2.284188] usbcore: registered new interface driver usbhid
[    2.284188] usbhid: USB HID core driver
[    2.285581] input: MOSART Semi. 2.4G Keyboard Mouse as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.1/1-1.1:1.0/0003:062A:5918.0001/input/input2
[    2.344479] hid-generic 0003:062A:5918.0001: input,hidraw0: USB HID v1.10 Keyboard [MOSART Semi. 2.4G Keyboard Mouse] on usb-0000:00:14.0-1.1/input0
[    2.344699] input: MOSART Semi. 2.4G Keyboard Mouse Mouse as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.1/1-1.1:1.1/0003:062A:5918.0002/input/input3
[    2.344890] input: MOSART Semi. 2.4G Keyboard Mouse Consumer Control as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.1/1-1.1:1.1/0003:062A:5918.0002/input/input4
[    2.404446] input: MOSART Semi. 2.4G Keyboard Mouse System Control as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.1/1-1.1:1.1/0003:062A:5918.0002/input/input5
[    2.404543] input: MOSART Semi. 2.4G Keyboard Mouse as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.1/1-1.1:1.1/0003:062A:5918.0002/input/input6
[    2.404744] hid-generic 0003:062A:5918.0002: input,hiddev0,hidraw1: USB HID v1.10 Mouse [MOSART Semi. 2.4G Keyboard Mouse] on usb-0000:00:14.0-1.1/input1
[    2.405450] input: Microsoft Surface Keyboard Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input8
[    2.464449] input: Microsoft Surface Keyboard Mouse as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input9
[    2.464655] input: Microsoft Surface Keyboard Consumer Control as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input10
[    2.464757] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input11
[    2.464859] input: Microsoft Surface Keyboard Touchpad as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input12
[    2.465026] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input13
[    2.465080] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input14
[    2.465138] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input15
[    2.465186] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input16
[    2.465235] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input17
[    2.465285] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input18
[    2.465331] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input19
[    2.465377] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input20
[    2.465421] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input21
[    2.465467] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input22
[    2.465513] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input23
[    2.465558] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input24
[    2.465604] input: Microsoft Surface Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input25
[    2.465651] input: Microsoft Surface Keyboard Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input26
[    2.465743] hid-generic 0003:045E:07CD.0003: input,hiddev1,hidraw2: USB HID v1.11 Keyboard [Microsoft Surface Keyboard] on usb-0000:00:14.0-1.4/input0
[    2.913095] input: Microsoft Surface Keyboard Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input42
[    2.972410] input: Microsoft Surface Keyboard Mouse as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input43
[    2.972586] input: Microsoft Surface Keyboard Consumer Control as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input44
[    2.972685] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input45
[    2.972776] input: Microsoft Surface Keyboard Touchpad as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input46
[    2.972894] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input47
[    2.972981] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input48
[    2.973033] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input49
[    2.973084] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input50
[    2.973135] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input51
[    2.973185] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input52
[    2.973234] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input53
[    2.973283] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input54
[    2.973330] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input55
[    2.973375] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input56
[    2.973422] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input57
[    2.973474] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input58
[    2.973520] input: Microsoft Surface Keyboard UNKNOWN as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input59
[    2.973582] input: Microsoft Surface Keyboard Keyboard as /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:045E:07CD.0003/input/input60
[    2.973683] hid-multitouch 0003:045E:07CD.0003: input,hiddev1,hidraw2: USB HID v1.11 Keyboard [Microsoft Surface Keyboard] on usb-0000:00:14.0-1.4/input0
[    3.167064] Console: switching to colour frame buffer device 375x125
[    3.192846] i915 0000:00:02.0: fb0: inteldrmfb frame buffer device

```

**Comment by battaglia01** [comment]
It could be. The `sudo kbd_mode -u` thing should only last until you restart, so perhaps try restarting and turning off your script and see if it works?

**Comment by calvous** [comment]
Same with Surface Pro 1796. With both 418 and 419. In 419 it was fixed in rc3. Using Arch  and it does it w/o jakedays patches applied.

**Comment by calvous** [comment]
I think it's fixed now with this kernel 4.18.9-1.

**Comment by battaglia01** [comment]
You've tried it on 4.18.9-1 and it works?

**Comment by jakeday** [comment]
Fixed in 4.18.11.
